postman: 7.0.7 -> 7.6.0

evanjs opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
https://www.getpostman.com/downloads/release-notes#changelog-linux-app-7.6.0

###### Things done
##### Major refactor
- This was primarily based on the GitKraken [expression](https://github.com/NixOS/nixpkgs/blob/7bc147a4966cfef5a27401c26ad1e8caddb9fc74/pkgs/applications/version-management/gitkraken/default.nix)
___
- Remove gnome2 (#39976)
  - Use pango instead of gnome2.pango
  - Remove gnome2.GConf
  - Remove gtk2-x11
- Add at-spi2-atk dependency
- Explicitly import packages rather than just pkgs or xorg
- Refactor patchelf to be more generic
- Use wrapGAppsHook
- Move libPath definition out of preFixup
- Remove dontPatchELF
- Remove unnecessary Postman binary, only use _Postman
___
###### Testing
- Open app
- Sent GET and POST requests and received responses
- Changed some settings
- Created a new collection
- Created a new request

Not experiencing any issues from the little I've tested so far.
___
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @xurei

jonringer commented on pkgs/development/web/postman/default.nix:
```diff
@@ -1,18 +1,53 @@
+, libXrender, libXtst, libXScrnSaver
+}:
+
+with stdenv.lib;
```
file wide "with's" should be discouraged. Since you're also receiving lib, you can just use "lib"
```suggestion
with stdenv.lib;
```

jonringer commented on pkgs/development/web/postman/default.nix:
```diff
@@ -1,18 +1,53 @@
-{ stdenv, lib, gnome2, fetchurl, pkgs, xorg, makeWrapper, makeDesktopItem }:
+{ stdenv, lib, fetchurl, pkgs, makeWrapper, makeDesktopItem, wrapGAppsHook
```
pkgs no longer used, stdenv.lib can be replaced with just lib
```suggestion
{ lib, fetchurl, makeWrapper, makeDesktopItem, wrapGAppsHook
```

jonringer commented on pkgs/development/web/postman/default.nix:
```diff
@@ -43,55 +81,21 @@ stdenv.mkDerivation rec {
   meta = with stdenv.lib; {
     homepage = https://www.getpostman.com;
     description = "API Development Environment";
     license = stdenv.lib.licenses.postman;
```
```suggestion
    license = licenses.postman;
```

jonringer commented on pkgs/development/web/postman/default.nix:
```diff
@@ -43,55 +81,21 @@ stdenv.mkDerivation rec {
+    popd
   '';
 
   meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```

jonringer commented on pkgs/development/web/postman/default.nix:
```diff
@@ -1,18 +1,53 @@
-  nativeBuildInputs = [ makeWrapper ];
-
-  dontPatchELF = true;
+  libPath = makeLibraryPath [
```
```suggestion
  libPath = lib.makeLibraryPath [
```

jonringer replied:
I would prefer this to be in a let binding, above the derivation (don't feel strongly about this), but it would avoid `$libPath` being set during the build

evanjs replied:
When you say build, do you mean the nix build, e.g. buildPhase?
Wasn't sure how it impacts this expression since we're not doing anything there right now.
`buildPhase = ":";`

jonringer replied:
im pretty sure all keys on a derivation get stringified to a variable for all phases. For example, saying `src = <some package>; unpackPhase = "tar xf $src"`, $src isn't populated because it's special, it's populated because it's an attribute on the derviation set.

jonringer reviewed (approved):
`nix-review` passes on NixOS
diff LGTM (A few nitpicks)
binary seems to work
leaf package

jonringer commented:
oh yea, stdenv.cc.cc, guess you will need stdenv :(. Good atempt though :)

you could try to do `ORIGIN=$(patch-elf --print-rpath $file); patch-elf --set-rpath "libPath:$ORIGIN" $file` but that's kind of ugly, but it would prevent you from destroying the default patching.

evanjs commented:
> oh yea, stdenv.cc.cc, guess you will need stdenv :(. Good atempt though :)
> 
> you could try to do `ORIGIN=$(patch-elf --print-rpath $file); patch-elf --set-rpath "libPath:$ORIGIN" $file` but that's kind of ugly, but it would prevent you from destroying the default patching.

Not sure if I'm reading this correctly, but it doesn't seem to work
``` diff
diff --git a/pkgs/development/web/postman/default.nix b/pkgs/development/web/postman/default.nix
index b56cdf6c811..15038bdfbc7 100644
--- a/pkgs/development/web/postman/default.nix
+++ b/pkgs/development/web/postman/default.nix
@@ -84,7 +84,7 @@ stdenv.mkDerivation rec {
     patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" _Postman
 
     for file in $(find . -type f \( -name \*.node -o -name _Postman -o -name \*.so\* \) ); do
-      patchelf --set-rpath ${libPath}:$out/share/postman $file || true
+      ORIGIN=$(patchelf --print-rpath $file); patchelf --set-rpath "libPath:$ORIGIN" $file || true
     done
     popd
   '';
```

`./result/bin/postman: error while loading shared libraries: libgtk-3.so.0: cannot open shared object file: No such file or directory`

jonringer commented:
> `./result/bin/postman: error while loading shared libraries: libgtk-3.so.0: cannot open shared object file: No such file or directory`
I'm not sure given the information, you can use `ldd` on a .so or elf to see where it did or didn't a given link. You may want to try `dontPatchELF = true;`, this will prevent the default phases from trying to reset rpath.

> Furthermore, is breaking the default patching even a concern with binary packages like this?
I find that most pre-built binaries are usually patched in this way :(.

evanjs commented:
Ah.  I think I just mistyped some things in my initial attempt based on your suggestions.
Had to move the makeLibraryPath to `postFixup` rather than `preFixup`, but everything seems to work now.

evanjs commented:
Anything else I can do to get this merged?

evanjs commented:
I have squashed the commits and separated the addition of myself to the list of maintainers.
I also bumped the package to 7.6.0.  Still works.

worldofpeace commented on pkgs/development/web/postman/default.nix:
```diff
@@ -26,12 +27,15 @@ stdenv.mkDerivation rec {
     categories = "Application;Development;";
   };
 
+  nativeBuildInputs = [ makeWrapper wrapGAppsHook ];
```
Where is `makeWrapper` used in the expression?

worldofpeace commented on pkgs/development/web/postman/default.nix:
```diff
@@ -1,20 +1,21 @@
-  dontPatchELF = true;
-
-  buildPhase = ":";   # nothing to build
+  dontBuild = true; # nothing to build
```
```suggestion
  dontBuild = true; # nothing to build
  dontConfigure = true;
```

worldofpeace reviewed (changes requested):
You need to add various libs to `buildInputs` for the application to be properly wrapped.
Else, when you launch a file chooser within the app, it will crash with " No GSettings schemas are installed on the system". Note that's in pure environments.

Perhaps add `libPath` to `buildInputs`.

jonringer commented on pkgs/development/web/postman/default.nix:
```diff
@@ -43,55 +48,52 @@ stdenv.mkDerivation rec {
+    pushd $out/share/postman
+    patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" _Postman
+    for file in $(find . -type f \( -name \*.node -o -name _Postman -o -name \*.so\* \) ); do
+      ORIGIN=$(patchelf --print-rpath $file); patchelf --set-rpath "${libPath}:$ORIGIN" $file || true
```
a lot going on in one line
```suggestion
      ORIGIN=$(patchelf --print-rpath $file)
      patchelf --set-rpath "${libPath}:$ORIGIN" $file || true
```

evanjs commented on pkgs/development/web/postman/default.nix:
```diff
@@ -1,20 +1,22 @@
-    sha256 = "47be1b955759520f3a2c7dcdecb85b4c52c38df717da294ba184f46f2058014a";
-    name = "${pname}-${version}.tar.gz";
+    sha256 = "sha256:03y82ydkj46l7dn35y944gnghbrrhc75y3yxdyidbh8fl3xvmlfv";
+    name = "${name}.tar.gz";
```
Unfortunately, the url differs from the actual file that is downloaded.
This is a workaround for that behavior. 
Without it, we get, e.g.:
```
unpacking source archive /nix/store/dvn4gjqcyj27h13lmfx13qf3pz0s2yn9-linux64
do not know how to unpack source archive /nix/store/dvn4gjqcyj27h13lmfx13qf3pz0s2yn9-linux64
```

jonringer replied:
that's fair, unfortunate that they have such an usual name. I'm fine with leavign it then

evanjs commented:
> You need to add various libs to `buildInputs` for the application to be properly wrapped.
> Else, when you launch a file chooser within the app, it will crash with " No GSettings schemas are installed on the system". Note that's in pure environments.
> 
> Perhaps add `libPath` to `buildInputs`.

Moving everything into `buildInputs` and constructing `libPath` off of that was one method I found worked well.
I wasn't sure how to otherwise add `libPath` to `buildInputs` without duplicating dependency references.

When attempting to remove `libPath` and utilize `autoPatchelfHook`, I get no abnormal build output, but the application dumps core before launching.

@worldofpeace Given the strange behavior when I try to use `autoPatchelfHook` and so on, do you have any thoughts on how the expression looks right now?

worldofpeace commented on pkgs/development/web/postman/default.nix:
```diff
@@ -43,55 +82,21 @@ stdenv.mkDerivation rec {
+    patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" _Postman
+    for file in $(find . -type f \( -name \*.node -o -name _Postman -o -name \*.so\* \) ); do
+      ORIGIN=$(patchelf --print-rpath $file); \
+      patchelf --set-rpath "${libPath}:$ORIGIN" $file
```
```suggestion
      patchelf --set-rpath "${stdenv.lib.makeLibraryPath buildInputs}:$ORIGIN" $file
```

worldofpeace commented on pkgs/development/web/postman/default.nix:
```diff
@@ -26,12 +28,49 @@ stdenv.mkDerivation rec {
+
+  nativeBuildInputs = [ wrapGAppsHook ];
+
+  libPath = stdenv.lib.makeLibraryPath buildInputs;
```
```suggestion
```

worldofpeace commented on pkgs/development/web/postman/default.nix:
```diff
@@ -1,20 +1,22 @@
-{ stdenv, lib, gnome2, fetchurl, pkgs, xorg, makeWrapper, makeDesktopItem }:
+{ stdenv, fetchurl, makeWrapper, makeDesktopItem, wrapGAppsHook
```
```suggestion
{ stdenv, fetchurl, makeDesktopItem, wrapGAppsHook
```

evanjs commented:
@worldofpeace Thank you!  This feels a lot less messy.
Is commit cleaning all that's left, now? (e.g. one for adding me to postman maintainers, one for the package updates)

worldofpeace reviewed (approved):
LGTM, verified file chooser worked with a cleared XDG_DATA_DIRS.

worldofpeace commented:
> @worldofpeace Thank you! This feels a lot less messy.
> Is commit cleaning all that's left, now? (e.g. one for adding me to postman maintainers, one for the package updates)

Yep, you can cleanup the git history and I'll merge :+1:

evanjs committed:
postman: 7.0.7 -> 7.6.0

- Remove gnome2 (#39976)
  - Use pango instead of gnome2.pango
  - Remove gnome2.GConf
  - Remove gtk2-x11
- Add at-spi2-atk dependency
- Explicitly import packages rather than just pkgs or xorg
- Refactor patchelf to be more generic
- Use wrapGAppsHook
  - As this app uses GTK3 for the UI, we need to use wrapGAppsHook
- Move libPath creation to postFixup
- Remove dontPatchELF
- Remove unnecessary Postman binary, only use _Postman
- Remove unnecessary makeWrapper
- Add dontConfigure
- Remove unnecessary lib 
- Move libPath inputs to buildInputs

evanjs committed:
postman: add evanjs to maintainers
